_posts/2020-10-21-explicit-term-inference-in-scala-3.md

```diff
@@ -0,0 +1,433 @@
+and on `implicit` parameters to propagate instances, which was a bit cryptic to beginners who were maybe already struggling with new concepts in functional programming.
+
+In Scala 3, this pattern becomes simpler thanks to the new syntax.
+Let's consider a simple typeclass such as `Show` which describes the capability of types to have a `String` representation.
+
+We first describe the interface that all instances of this typeclass should implement:
+
+```scala
+// Show.scala
+trait Show[A]:
+  extension (a: A) def show: String
+```
+We can then add a companion object to provide a couple of auxiliary methods and instances
+that make instance creation less tedious:
+```scala
+// Show.scala
+object Show:
+  def from[A](f: A => String): Show[A] = new Show[A]:
+    extension (a: A) def show: String = f(a)
+
+  given[A: Show] as Show[List[A]] = new Show[List[A]]:
+    extension (ls: List[A]) def show: String = ls.map(_.show).mkString(",")
+```
+
+To use it we need to import the interface and define an instance:
+```scala
+// Main.scala
+import Show.{_, given}
+
+case class Mountain(name: String, height: Int)
+
+given Show[Mountain] =
+  Show.from((m: Mountain) => s"The ${m.name} is ${m.height} meters high")
+
+@main def main =
+  val mountains = List(Mountain("Mont Blanc", 4808), Mountain("Matterhorn", 4478))
+  println(mountains.show)
+```
+As the extension method is defined inside the trait, there is no additional import statement required contrarily to what used to happen in Scala 2, as you will see in the following section.
+
+### How was this done in Scala 2?
+
+In Scala 2 there was more boilerplate code involved. It all starts with defining the interface:
+
+```scala
+// Show.scala
+trait Show[A] {
+  def show(a: A): String
+}
+```
+The second step was to define the extension method which relies on two implicits:
+```scala
+// ShowOps.scala
+object ShowOps {
+  implicit class showOps[A](in: A) extends AnyVal {
+    def show(implicit instance: Show[A]): String =
+      instance.show(in)
+  }
+}
+```
+And finally we can define the auxiliary methods in the companion object:
+```scala
+// Show.scala
+object Show {
+  def from[A](f: A => String): Show[A] = new Show[A] {
+    def show(a: A): String = f(a)
+  }
+
+  // Either this or import ShowOps and use context bound
+  implicit def showList[A](ls: List[A])(implicit instance: Show[A]) =
+    ls.map(instance.show).mkList(",")
+}
+```
+which brings us to the main definition:
+```scala
+// Main.scala
+import Show._
+import ShowOps._
+
+case class Mountain(name: String, height: Int)
+
+object Main extends App {
+  implicit val mountainShow: Show[Mountain] =
+    Show.fromFunction((m: Mountain) => s"The ${m.name} is ${m.height} meters high")
+
+  val mountains = List(Mountain("Mont Blanc", 4808), Mountain("Matterhon", 4478))
+  println(mountains.show)
+}
+```
+I believe that this example shows how `implicit` was used to achieve different goals
+and, in doing so, used to be more confusing:
```

Comment:
Due to the first part of the sentence when "used to" is actually about *using* implicits, I read the second part of as "people used implicits to be more confusing" (i.e., people were doing their best to be confusing), and not as "implicits were more confusing back then". Consider:
```suggestion
I believe that this example shows how `implicit`s were used to achieve different goals
and, in doing so, how they used to be more confusing:
```
which creates the right symmetry around "how" rather than around "used to".